pkg/reconciler/common/read_releases.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+
+	"golang.org/x/mod/semver"
+)
+
+// RetrieveManifestPath returns the manifest path for Knative based a provided version and component
+func RetrieveManifestPath(version, kcomponent string) string {
+	koDataDir := os.Getenv("KO_DATA_PATH")
+	return filepath.Join(koDataDir, kcomponent, version)
+}
+
+// ListReleases returns the all the available release versions available under kodata directory for Knative component.
+func ListReleases(kComponent string) []string {
+	releaseTags := []string{}
+	// List all the directories available under kodata
+	koDataDir := os.Getenv("KO_DATA_PATH")
+	pathname := filepath.Join(koDataDir, kComponent)
+	fileList, err := ioutil.ReadDir(pathname)
+	if err != nil {
+		panic(err)
+	}
+	for _, file := range fileList {
+		name := path.Join(pathname, file.Name())
+		pathDirOrFile, err := os.Stat(name)
+		if err != nil {
+			panic(err)
+		}
+		if pathDirOrFile.IsDir() {
+			releaseTags = append(releaseTags, file.Name())
+		}
+	}
+	if len(releaseTags) == 0 {
+		panic(fmt.Errorf("unable to find any version number for %s", kComponent))
+	}
+
+	// This function makes sure the versions are sorted in a descending order.
+	sort.Slice(releaseTags, func(i, j int) bool {
+		// Add the prefix "v" in front of the version
+		versionI := fmt.Sprintf("v%s", releaseTags[i])
+		versionJ := fmt.Sprintf("v%s", releaseTags[j])
+		return semver.Compare(versionI, versionJ) == 1
```

Comment:
I didn't realize that library was so strict about the "v" prefix. Would it make things simpler to include the "v" in the directory name?

Reply:
We do not put the "v" in the status.version. We need to convert it either here, or checking with the status.version or even spec.version.
There are potentially three places to consider the name of the versions.
```
Dir names
status.version
even spec.version.

Do we consider changing the status.version by adding "v"?

Reply:
I don't have a strong opinion. I'm for whatever makes the code (and the user's life) simpler. I honestly thought the semver spec treated the 'v' as optional -- probably got that idea from Gopkg.toml.

Reply:
SemVer lib takes the "v" as part of the MAJOR. It does not conform to the definition and the example here: https://semver.org/(I do not see `v` in front of all the versions). I tend to keep the format as major.minor.patch, without the letter `v` for our code.
We only add `v`, when comparing two versions.

Reply:
I can create a function called `func isFormerVersionLater(former, latter string) bool`, to compare two versions. It adds `v`, if it is missing.